### PR TITLE
PDEP6 implementation pt 1: block.setitem, block.putmask

### PR DESCRIPTION
**Pull request by MarcoGorelli**
This is pt 1. More changes would be needed to fully complete PDEP6 (`eablock.setitem` and `eablock.putmask` for a start), but I'd suggest that we do this in stages. This PR is already pretty big

some of this can be factored out, I'll do that first

pt2: https://github.com/pandas-dev/pandas/pull/53405

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
@jbrockmendel gentle ping, just to ask if you think this would be the right direction / the right place to raise the warning. I think there would need to be at least 4 such PRs to address the initial scope of PDEP6

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -945,6 +947,13 @@ def setitem(self, indexer, value) -> Block:
             casted = np_can_hold_element(values.dtype, value)
         except LossySetitemError:
             # current dtype cannot store value, coerce to common dtype
+            warnings.warn(
+                f"Setting an item of incompatible dtype is deprecated "
+                "and will raise in a future error of pandas. "
+                f"Value {value} has dtype incompatible with {values.dtype}",
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:953]: "... explicitly cast to whatever first"

**Comment by jbrockmendel** [comment]
> the right place to raise the warning

Looks right to me

File: pandas/tests/series/indexing/test_setitem.py [hunk 2]
```diff
@@ -536,22 +557,30 @@ def test_setitem_keep_precision(self, any_numeric_ea_dtype):
...
+            (np.nan, NA, "Float64", "Float64", 1, None),
+            (np.nan, NA, "Float64", "Float64", 2, None),
+            (np.nan, np.nan, "int64", "float64", 1, FutureWarning),
+            (np.nan, np.nan, "int64", "float64", 2, None),
         ],
     )
-    def test_setitem_enlarge_with_na(self, na, target_na, dtype, target_dtype, indexer):
+    def test_setitem_enlarge_with_na(
+        self, na, target_na, dtype, target_dtype, indexer, warn
+    ):
```
Comment (MarcoGorelli) [inline pandas/tests/series/indexing/test_setitem.py:579]: note how here, the warnings are only ever raised when there is no enlargement  (i.e. when the indexer is `1`, which is within-bounds)

**Comment by MarcoGorelli** [comment]
🤔 this is still raising a warning for
```python
s = pd.Series([1, 2, 3])
s.update(pd.Series([4, np.nan, 6]))
print(s)
```
, looking into it

File: pandas/core/internals/blocks.py [hunk 3]
```diff
@@ -412,7 +414,7 @@ def split_and_operate(self, func, *args, **kwargs) -> list[Block]:
     # Up/Down-casting
 
     @final
-    def coerce_to_target_dtype(self, other) -> Block:
+    def coerce_to_target_dtype(self, other, warn_on_upcast: bool = False) -> Block:
```
Comment (MarcoGorelli) [inline pandas/core/internals/blocks.py:417]: this could be changed to `raise_on_upcast` when the deprecation is enforced

File: doc/source/user_guide/categorical.rst [hunk 4]
```diff
@@ -779,6 +779,7 @@ Setting values by assigning categorical data will also check that the ``categori
 Assigning a ``Categorical`` to parts of a column of other types will use the values:
 
 .. ipython:: python
+    :okwarning:
```
[the remark this answers is not in the document]
Comment (mroeschke) [reply doc/source/user_guide/categorical.rst:782]: Could you convert these to code blocks?

**Review by mroeschke — COMMENTED** [review]
Do we have tests where this will warn when setting an EA type value?

**Review by phofl — COMMENTED** [review]
Implementation itself looks good. I guess there is no way of reducing the test diff without changing the tests?

**Comment by MarcoGorelli** [comment]
yeah I've factored out into separate PRs what I could, but I'll see if I can do more

File: doc/source/user_guide/categorical.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (MarcoGorelli) [reply doc/source/user_guide/categorical.rst:782]: ok for old whatsnew notes, but wouldn't changing this to a code block in the user guide lead us to forget about it and have incorrect examples in the future?

[the remark this answers is not in the document]
Comment (mroeschke) [reply doc/source/user_guide/categorical.rst:782]: Ah good point. Yeah fine to leave then

File: pandas/core/internals/blocks.py [hunk 5]
```diff
@@ -1043,7 +1053,7 @@ def setitem(self, indexer, value, using_cow: bool = False) -> Block:
             casted = np_can_hold_element(values.dtype, value)
         except LossySetitemError:
             # current dtype cannot store value, coerce to common dtype
-            nb = self.coerce_to_target_dtype(value)
+            nb = self.coerce_to_target_dtype(value, warn_on_upcast=True)
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:1056]: looks like this is just in Block. dont we need the same thing in EABackedBlock?

Comment (MarcoGorelli) [reply pandas/core/internals/blocks.py:1056]: yup! OK to do that in a separate PR to keep the files changed down?

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:1056]: sure

**Comment by MarcoGorelli** [comment]
gentle ping

image

File: doc/source/whatsnew/v0.21.0.rst [hunk 6]
```diff
@@ -425,13 +425,12 @@ Note that this also changes the sum of an empty ``Series``. Previously this alwa
...
+but for consistency with the all-NaN case, this was changed to return 0 as well:
 
-.. ipython:: python
-   :okwarning:
-
-   pd.Series([]).sum()
+.. code-block:: ipython
 
+   In [2]: pd.Series([]).sum()
+   Out[2]: 0
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v0.21.0.rst:433]: this is unrelated to this PR, just incorrect AFAICT. (mostly a note-to-self)

File: pandas/core/internals/blocks.py [hunk 7]
```diff
@@ -423,7 +425,15 @@ def coerce_to_target_dtype(self, other) -> Block:
         and will receive the same block
         """
         new_dtype = find_result_type(self.values, other)
-
+        if warn_on_upcast and self.dtype != new_dtype:
```
Comment (jbrockmendel) [inline pandas/core/internals/blocks.py:428]: if a user ever gets here with `new_dtype=self.dtype` then something has gone wrong and they are going to get a RecursionError.  we should probably add a check for that and raise with a "please open a bug report"

File: pandas/tests/frame/indexing/test_indexing.py [hunk 8]
```diff
@@ -1332,12 +1338,22 @@ def test_loc_expand_empty_frame_keep_midx_names(self):
...
-    @pytest.mark.parametrize("val", ["x", 1])
-    @pytest.mark.parametrize("idxr", ["a", ["a"]])
-    def test_loc_setitem_rhs_frame(self, idxr, val):
+    @pytest.mark.parametrize(
+        "val, idxr, warn",
+        [
+            ("x", "a", None),
+            ("x", ["a"], None),
+            (1, "a", None),
+            (1, ["a"], FutureWarning),
```
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_indexing.py:1347]: this seems weird to me.  the others also cast, right?

Comment (MarcoGorelli) [reply pandas/tests/frame/indexing/test_indexing.py:1347]: agree that it's weird - they don't because the RHS of the assignment is a `DataFrame`, rather than a single value or a Series, so it goes down a different path. I've added comments to indicate this is todo

File: pandas/tests/series/indexing/test_indexing.py [hunk 9]
```diff
@@ -188,12 +195,12 @@ def test_setitem_ambiguous_keyerror(indexer_sl):
 
 def test_setitem(datetime_series):
     datetime_series[datetime_series.index[5]] = np.NaN
-    datetime_series[[1, 2, 17]] = np.NaN
-    datetime_series[6] = np.NaN
-    assert np.isnan(datetime_series[6])
-    assert np.isnan(datetime_series[2])
+    datetime_series.iloc[[1, 2, 17]] = np.NaN
```
Comment (jbrockmendel) [inline pandas/tests/series/indexing/test_indexing.py]: isnt this test specifically for `__setitem__`?

File: pandas/tests/series/indexing/test_indexing.py [hunk 10]
```diff
@@ -292,9 +299,9 @@ def test_underlying_data_conversion(using_copy_on_write):
 
 
 def test_preserve_refs(datetime_series):
-    seq = datetime_series[[5, 10, 15]]
-    seq[1] = np.NaN
-    assert not np.isnan(datetime_series[10])
+    seq = datetime_series.iloc[[5, 10, 15]]
+    seq.iloc[1] = np.NaN
```
Comment (jbrockmendel) [inline pandas/tests/series/indexing/test_indexing.py]: same here

File: pandas/tests/series/indexing/test_indexing.py (hunk 9 above)
Comment (MarcoGorelli) [reply pandas/tests/series/indexing/test_indexing.py]: this is an outdated file, I didn't modify it - I presume you're commenting on the merge commit?

File: pandas/tests/series/indexing/test_indexing.py (hunk 10 above)
Comment (MarcoGorelli) [reply pandas/tests/series/indexing/test_indexing.py]: I didn't modify this one either - I presume you're commenting on the merge commit?

File: pandas/core/internals/blocks.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:419]: this would also cover 'shift' (xref #53802).  I don't recall that being a part of PDEP6.  Am I remembering wrong?

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:419]: oh i see there's a new kwd, never mind

File: pandas/tests/series/methods/test_convert_dtypes.py [hunk 11]
```diff
@@ -206,7 +206,20 @@ def test_convert_dtypes(
         # Test that it is a copy
         copy = series.copy(deep=True)
 
-        result[result.notna()] = np.nan
+        if result.notna().sum() > 0 and result.dtype in [
```
Comment (jbrockmendel) [inline pandas/tests/series/methods/test_convert_dtypes.py:209]: doesnt have to be for this PR, but long-term should we also warn/raise in the no-op case?

File: pandas/tests/series/indexing/test_where.py [hunk 12]
```diff
@@ -38,23 +38,24 @@ def test_where_unsafe_float(float_numpy_dtype):
...
-        (np.int32, np.float64),
-        (np.int64, np.float64),
-        (np.float32, np.float32),
-        (np.float64, np.float64),
+        (np.int8, np.float64, FutureWarning),
+        (np.int16, np.float64, FutureWarning),
+        (np.int32, np.float64, FutureWarning),
+        (np.int64, np.float64, FutureWarning),
+        (np.float32, np.float32, None),
+        (np.float64, np.float64, None),
```
Comment (jbrockmendel) [inline pandas/tests/series/indexing/test_where.py:48]: can the `warn` variable be set inside the function?  i think that can make the reasoning for how it is defined clearer (e.g `warn = None if expected_dtype.kind == dtype.kind == "f" else FutureWarning`)

**Comment by jbrockmendel** [comment]
9/21 files viewed!  Getting there...

File: pandas/tests/series/methods/test_convert_dtypes.py (hunk 11 above)
Comment (MarcoGorelli) [reply pandas/tests/series/methods/test_convert_dtypes.py:209]: as in, warn if
```python
ser[empty_mask] = incompatible_value
```
?

I wouldn't think that would be necessary, but we can discuss it

File: pandas/tests/series/methods/test_update.py [hunk 13]
```diff
@@ -39,34 +39,35 @@ def test_update(self, using_copy_on_write):
...
+            ([61.1, 63.1], object, Series([10, 61.1, 12], dtype=object), None),
             # other is object, cannot be cast
-            ([(61,), (63,)], "int32", Series([10, (61,), 12])),
-            ([(61,), (63,)], "int64", Series([10, (61,), 12])),
-            ([(61,), (63,)], float, Series([10.0, (61,), 12.0])),
-            ([(61,), (63,)], object, Series([10, (61,), 12])),
+            ([(61,), (63,)], "int32", Series([10, (61,), 12]), FutureWarning),
+            ([(61,), (63,)], "int64", Series([10, (61,), 12]), FutureWarning),
+            ([(61,), (63,)], float, Series([10.0, (61,), 12.0]), FutureWarning),
+            ([(61,), (63,)], object, Series([10, (61,), 12]), None),
```
Comment (jbrockmendel) [inline pandas/tests/series/methods/test_update.py:63]: is there a nice way of defining `warn` inside the test function e.g.

```
warn = None
if some_intutiive_condition:
    warn = FutureWarning
```

Comment (MarcoGorelli) [reply pandas/tests/series/methods/test_update.py:63]: to be honest I generally prefer to avoid logic (like if-then) statements in tests https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html

can do if you prefer, just explaining why I intentionally did it this way

Comment (jbrockmendel) [reply pandas/tests/series/methods/test_update.py:63]: ill give that a read.  my thought is that it is often not obvious why some cases have FutureWarning which can be a real problem when a test can be wrong(ish) eg https://github.com/pandas-dev/pandas/pull/50626#discussion_r1199642076

File: pandas/tests/series/indexing/test_indexing.py [hunk 14]
```diff
@@ -444,3 +445,61 @@ def test_setitem_dict_and_set_disallowed_multiindex(self, key):
...
+            ser = orig_ser.copy()
+
+        with tm.assert_produces_warning(FutureWarning, match=msg):
+            ser.loc[indexer] = invalid
+            ser = orig_ser.copy()
+
+        # note: commented-out in the EA case too
+        # FIXME: don't leave commented-out
+        # with tm.assert_produces_warning(FutureWarning, match=msg):
+        #     ser[:] = invalid
```
Comment (jbrockmendel) [inline pandas/tests/series/indexing/test_indexing.py:474]: this doesnt match the commented-out code in the masked tests.  any chance it can be un-commented?

File: pandas/tests/frame/test_constructors.py [hunk 15]
```diff
@@ -2548,8 +2555,20 @@ def check_views(c_only: bool = False):
...
         # TODO: most of the rest of this test belongs in indexing tests
-        df.iloc[0, 0] = 0
-        df.iloc[0, 1] = 0
+        if (
+            is_float_dtype(df.dtypes.iloc[0])
+            or is_integer_dtype(df.dtypes.iloc[0])
+            or is_complex_dtype(df.dtypes.iloc[0])
+            or is_object_dtype(df.dtypes.iloc[0])
+            or is_datetime64_dtype(df.dtypes.iloc[0])  # TODO this one should warn
+            or is_timedelta64_dtype(df.dtypes.iloc[0])  # TODO this one should warn
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_constructors.py:2564]: are these TODOs things you want to do for this PR or follow-ups?

File: pandas/tests/frame/test_constructors.py [hunk 16]
```diff
@@ -2548,8 +2555,20 @@ def check_views(c_only: bool = False):
             check_views()
 
         # TODO: most of the rest of this test belongs in indexing tests
-        df.iloc[0, 0] = 0
-        df.iloc[0, 1] = 0
+        if (
+            is_float_dtype(df.dtypes.iloc[0])
```
Comment (jbrockmendel) [inline pandas/tests/frame/test_constructors.py:2559]: NBD but the suggested usage for these checks is now `lib.is_np_dtype(df.dtypes.iloc[0], "fciuomM")`

File: pandas/tests/frame/indexing/test_indexing.py [hunk 17]
```diff
@@ -1851,3 +1870,54 @@ def test_setitem_dict_and_set_disallowed_multiindex(self, key):
         )
         with pytest.raises(TypeError, match="as an indexer is not supported"):
             df.loc[key] = 1
+
+
+class TestSetitemValidation:
```
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_indexing.py:1875]: is there a viable way to share this with the analogous copy/pasted Series tests?

File: pandas/tests/frame/indexing/test_where.py [hunk 18]
```diff
@@ -189,6 +189,8 @@ def _check_set(df, cond, check_dtypes=True):
             with pytest.raises(TypeError, match=msg):
                 df > 0
             return
+        if df is mixed_int_frame:
```
Comment (jbrockmendel) [inline pandas/tests/frame/indexing/test_where.py:192]: can this be reached?

File: pandas/tests/series/indexing/test_setitem.py [hunk 19]
```diff
@@ -191,11 +191,14 @@ def test_setitem_series_object_dtype(self, indexer, ser_index):
...
 
-    @pytest.mark.parametrize("index, exp_value", [(0, 42), (1, np.nan)])
-    def test_setitem_series(self, index, exp_value):
+    @pytest.mark.parametrize(
+        "index, exp_value, warn", [(0, 42, None), (1, np.nan, FutureWarning)]
+    )
+    def test_setitem_series(self, index, exp_value, warn):
         # GH#38303
         ser = Series([0, 0])
-        ser.loc[0] = Series([42], index=[index])
```
Comment (jbrockmendel) [inline pandas/tests/series/indexing/test_setitem.py:198]: unrelatedly, this looks sketchy since the RHS is a sequence and the LHS is a scalar

**Comment by jbrockmendel** [comment]
small comments, otherwise LGTM

File: pandas/core/internals/blocks.py [hunk 20]
```diff
@@ -424,7 +426,21 @@ def coerce_to_target_dtype(self, other) -> Block:
         and will receive the same block
         """
         new_dtype = find_result_type(self.values, other)
-
+        if warn_on_upcast:
+            warnings.warn(
+                f"Setting an item of incompatible dtype is deprecated "
+                "and will raise in a future error of pandas. "
+                f"Value '{other}' has dtype incompatible with {self.dtype}, "
+                "please explicitly cast to a compatible dtype first.",
```
Comment (mroeschke) [inline pandas/core/internals/blocks.py:434]: Thinking aloud. It would be difficult to suggest which dtype to cast? I guess `object` would be safest though

Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:434]: in practice i think it will almost always be x->object or int->float

File: pandas/tests/frame/indexing/test_where.py (hunk 18 above)
Comment (MarcoGorelli) [reply pandas/tests/frame/indexing/test_where.py:192]: yup: `pandas/tests/frame/indexing/test_where.py::TestDataFrameIndexingWhere::test_where_set[mixed_int]`

File: pandas/tests/frame/test_constructors.py (hunk 15 above)
Comment (MarcoGorelli) [reply pandas/tests/frame/test_constructors.py:2564]: this one addresses them: https://github.com/pandas-dev/pandas/pull/53405

File: pandas/tests/frame/test_constructors.py (hunk 16 above)
Comment (MarcoGorelli) [reply pandas/tests/frame/test_constructors.py:2559]: nice, that simplifies things, thanks!

File: pandas/tests/frame/indexing/test_indexing.py (hunk 17 above)
Comment (MarcoGorelli) [reply pandas/tests/frame/indexing/test_indexing.py:1875]: maybe, but I can't think of anything that wouldn't overcomplicate it - any suggestions?

File: pandas/core/internals/blocks.py (hunk 20 above)
Comment (MarcoGorelli) [reply pandas/core/internals/blocks.py:460]: maybe - but I'd suggest doing this separately if it's OK, this has already been open since January 😄 reckon we can get this in for 2.1? It should be pretty quick to rebase #53405, can do that right after this one

Comment (mroeschke) [reply pandas/core/internals/blocks.py:460]: Yep this would be fine as a follow up

**Comment by mroeschke** [comment]
Looks pretty good to me too. 

Should this have a whatsnew note for 2.1? And is it okay if not all of PDEP6 changes don't make it into 2.1?

File: pandas/core/internals/blocks.py (hunk 20 above)
Comment (jbrockmendel) [reply pandas/core/internals/blocks.py:460]: Works for me

File: pandas/tests/frame/indexing/test_indexing.py (hunk 17 above)
Comment (jbrockmendel) [reply pandas/tests/frame/indexing/test_indexing.py:1875]: Not worth worrying about

**Comment by MarcoGorelli** [comment]
Thanks!

I think ideally this and https://github.com/pandas-dev/pandas/pull/53405 should go together, either both in 2.1 or both in 2.2

Whatsnew - #54014 introduces a section for notable deprecations, shall I add it to that once that's merged?
Or I can add it as part of #53405

**Comment by mroeschke** [comment]
Probably best if the whatsnew is added in https://github.com/pandas-dev/pandas/pull/53405 then. Is that PR pretty close?

**Comment by MarcoGorelli** [comment]
yeah it should be ready, I just need to rebase it after this one - but it doesn't look like there's too many conflicts to resolve fortunately

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @MarcoGorelli

**Comment by jbrockmendel** [comment]
nice!
